cosmwasm/osmosis-tokenfactory-token-minter/src/contract.rs

```diff
@@ -0,0 +1,315 @@
+    prost::Name,
+};
+
+pub const DEFAULT_DECIMALS: u8 = 6;
+
+use crate::{
+    error::Error,
+    state::{ADMIN, TOKEN_ADMIN},
+};
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _: Env,
+    info: MessageInfo,
+    msg: TokenMinterInitMsg,
+) -> Result<Response, Error> {
+    let TokenMinterInitMsg::OsmosisTokenFactory { zkgm_admin } = msg else {
+        return Err(Error::InvalidMinterConfig);
+    };
+    ADMIN.save(deps.storage, &info.sender)?;
+    TOKEN_ADMIN.save(deps.storage, &zkgm_admin)?;
+    Ok(Response::default())
+}
+
+#[cw_serde]
+pub struct MigrateMsg {}
+
+#[entry_point]
+pub fn migrate(_: DepsMut, _: Env, _: MigrateMsg) -> StdResult<Response> {
+    Ok(Response::new())
+}
+
+fn deconstruct_factory_denom<'a>(env: &Env, denom: &'a str) -> Result<&'a str, Error> {
+    let denom_parts = denom
+        .split_once('/')
+        .and_then(|(a, b)| b.split_once('/').map(|(b, c)| (a, b, c)));
+
+    match denom_parts {
+        Some(("factory", addr, subdenom)) if addr == env.contract.address.as_str() => Ok(subdenom),
+        _ => Err(Error::InvalidDenom(denom.to_string())),
+    }
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<TokenFactoryMsg>, Error> {
+    if info.sender != ADMIN.load(deps.storage)? {
+        return Err(Error::OnlyAdmin);
+    }
+
+    let resp = match msg {
+        ExecuteMsg::Wrapped(msg) => {
+            let msgs = match msg {
+                WrappedTokenMsg::CreateDenom {
+                    subdenom: denom,
+                    metadata,
+                    ..
+                } => {
+                    let subdenom = deconstruct_factory_denom(&env, &denom)?;
+
+                    vec![
+                        CosmosMsg::Custom(TokenFactoryMsg::CreateDenom {
+                            subdenom: subdenom.to_owned(),
+                        }),
+                        #[allow(deprecated)]
+                        CosmosMsg::Stargate {
+                            type_url: MsgSetDenomMetadata::type_url(),
+                            value: MsgSetDenomMetadata {
+                                sender: env.contract.address.to_string(),
+                                metadata: Some(protos::cosmos::bank::v1beta1::Metadata {
+                                    description: "".to_string(),
+                                    denom_units: vec![
+                                        DenomUnit {
+                                            denom: denom.clone(),
+                                            exponent: 0,
+                                            aliases: vec![metadata.symbol.clone()],
+                                        },
+                                        DenomUnit {
+                                            denom: metadata.symbol.clone(),
+                                            exponent: metadata.decimals.into(),
```

Comment:
please confirm this is correct @aeryz, it was hardcoded to 18 previously

Reply:
We changed it to 6 few weeks ago when you were sleeping lol. The reasoning is the `DEFAULT_DECIMALS` is only used when you can't find the decimal info for the local token. And cosmos native tokens are usually 6 decimals.